File: customer.erl
%%%-------------------------------------------------------------------
%%% @author vanduong
%%% @copyright (C) 2019, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 11. Jun 2019 6:03 PM
%%%-------------------------------------------------------------------
-module(customer).
-author("vanduong").

%% API
-export([startCustomer/4]).

toString(Val)->
  lists:flatten(io_lib:format("~p", [Val])).

%% @doc starts a customer with given name, inital amount and bank list.
%% @param  ParentID : parent process id.
%% @param Name : customer name.
%% @param Amount : initial amount
%% @param BankList : list of banks
startCustomer(ParentID, Name, Amount, BankList)->
  timer:sleep(1000), % sleep 100 ms to make sure other processes are up
  borrow(ParentID, Name, Amount, BankList).

%% @doc randomly selects a bank from bank list and returns it.
%% @param BankList : list of banks
%% @returns a bank tuple or empty of list is empty
getRandomBank(BankList)->
  case BankList of
    [] -> {};
    _->
    lists:nth(rand:uniform(length(BankList)), BankList)
  end.

%% @doc randomly returns an amount between 1 and give amount.
%% @param Total : amount
%% @returns an amount
calculateLoan(Total)->
  if
    Total == 0->
      0;
    Total > 50->
      rand:uniform(50);
    true-> %match with anything
      rand:uniform(Total)
  end.

%% @doc checks if customer can/need to borrow more money, the process terminates if not.
%% @param ParentID : parent process id
%% @param Name : customer name
%% @param Total : current amount needs to loan
%% @param BankList : list of banks
checkValidity(ParentID, Name, Total, BankList)->
  if
    Total == 0-> %no need to borrow anymore
      ParentID ! {self(), {satisfied, Name, Total}},
      exit(normal);
    length(BankList) == 0-> %empty BankList, cannot borrow anymore
      ParentID ! {self(), {unsatisfied, Name, Total}},
      exit(normal);
    true->ok
  end.

%% @doc proceeds borrowing, updates total loan amount and bank list if needed.
%% @param ParentID : parent id process
%% @param Name : customer name
%% @param Total : current loan amount
%% @param BankLis : list of banks
borrow(ParentID, Name, Total, BankList) ->
  timer:sleep(rand:uniform(100) + 10), %sleep from 10-100 ms
  checkValidity(ParentID, Name, Total, BankList),
  {Bank, _} = getRandomBank(BankList),
  Loan = calculateLoan(Total),
  %send a message includes id, Name, amount to borrow, Bank
  ParentID ! {self(), {ask, Name, Loan, Bank}},
  %ParentID !  {toString(Name) ++ " started with " ++ toString(Amount) ++ " with bank list " ++ toString(BankList)}.
  {Banker, Amount} = get_feedback(),
  if
    Amount > 0->
      NewTotal = Total - Amount,
      borrow(ParentID, Name, NewTotal, BankList);
    true->
      %remove the bank from BankList
      borrow(ParentID, Name, Total, lists:keydelete(Banker,1, BankList))
  end.

%% @doc listens to message sent to this customer.
get_feedback() ->
  %if loan is granted, deduct loan amount from Amount, else remove bank from BankList
  receive
    {approved, {Banker, Loan}} ->
      {Banker, Loan};
    {rejected, {Banker, Loan}} ->
      {Banker, 0}

%%      get_feedback(ParentID, Name)
%%  after 2000 -> ParentID ! {"~w stops listening \n", [Name]}
  end.

File: Customer.java
/**
 * 
 */
package main;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * The Class Customer.
 *
 * @author vanduong
 */
public class Customer implements Runnable{
	
	/** The customer. */
	private String customer;
	
	/** The total. */
	private int total;
	
	/** The bank list. */
	private List<String> bankList;
	
	/**
	 * Instantiates a new customer.
	 *
	 * @param customer the customer
	 * @param total the total
	 * @param list the list
	 */
	public Customer(String customer, int total, List<String> list) {
		super();
		this.customer = customer;
		this.total = total;
		this.bankList = list;
	}

	/* (non-Javadoc)
	 * @see java.lang.Runnable#run()
	 */
	@Override
	public void run() {
		try {
			Thread.sleep(1000);//sleep 1000ms to make sure other threads are up
//			Money.cus2main.put(new Message("ask", customer, "rbc", 200));
//			System.out.println(customer + "received message from " + Money.main2cus.take().bank);
			borrow();
			
		} catch (InterruptedException e) {
			System.out.println("Something went wrong in customer");
			e.printStackTrace();
		}
		
	}

	/**
	 * Borrow.
	 *
	 * @throws InterruptedException the interrupted exception
	 */
	private void borrow() throws InterruptedException {
		while(!bankList.isEmpty() && total != 0) {
			//get random bank
			String bank = bankList.get(ThreadLocalRandom.current().nextInt(0, bankList.size()));	
			//get random loan
			int loan = 0;
			if(this.total > 50) {
				loan = ThreadLocalRandom.current().nextInt(1, 50 + 1);
			}else {
				loan = ThreadLocalRandom.current().nextInt(1, total + 1);
			}
			//send message to Money
			Money.cus2main.put(new Message("ask", customer, bank, loan));
			Message response = null;
			//check for response
			while(true) {
				response = Money.main2cus.take();
				if(response.customer.equalsIgnoreCase(this.customer)) {
					break;
				}else {
					Money.main2cus.put(response);//put back to the queue if not for this customer
				}
			}
			//check response detail
			if(response.label.equalsIgnoreCase("ok")) {
				//deduct total amount
				this.total -= response.amount;
			}else if(response.label.equalsIgnoreCase("no")) {
				if(bankList.contains(response.bank)) {
					//remove bank from bank list
					this.bankList.remove(bankList.indexOf(response.bank));
				}
				
			}
			Thread.sleep(ThreadLocalRandom.current().nextInt(10, 100 + 1));//sleep between loans

		}
		if(total == 0) {
			Money.cus2main.put(new Message("satisfied", customer, null, this.total));
		}else if(bankList.isEmpty()) {
			Money.cus2main.put(new Message("unsatisfied", customer, null, this.total));
		}
		
	}
	

}
